LCB Compile modules as c for iOS static linking

montegoulding opened the pull request.

montegoulding commented on toolchain/lc-compile/src/emit.cpp:
```diff
@@ -1109,10 +1111,41 @@ void EmitVariableDefinition(intptr_t p_index, PositionRef p_position, NameRef p_
+        }
+        else
+        {
+            MCAutoStringRefAsCString t_symbol_c_string;
```
@runrevmark I guess at this point we should be building a list of unique `(*t_info)->c.library` and writing to stdout at the end for deps

montegoulding committed:
[[ Module Lcext ]] Generate module lcext for ini example

montegoulding committed:
[[ Module Lcext ]] Update iOS Standalone builder handle module lcexts

This patch updates the standalone builder to link module lcexts and include the
module.lcm for them which is just the name of the module. Most noteworthy is the
removal of deadstripping by the standalone builder. All lcexts being linked together
should be dead stripped when built. This change allows us to link lcexts without
forcing the `__libinfoptr_<name>` which is the method used to link externals and
is not in the module lcext.

montegoulding committed:
[[ Module Lcext ]] Generate module lcext for timezone library

montegoulding commented on ide-support/revsaveasiosstandalone.livecodescript:
```diff
@@ -333,21 +340,25 @@ private command revSaveAsMobileStandaloneMain pStack, pAppBundle, pTarget, pSett
    -- MM-2012-09-18: Use the new copied externals array instead of the copy files.
    --
    local tDeployExternals, tLibraryMappings
+   put pSettings["ios,library"] into tLibraryMappings 
```
This needs to be back-ported to develop-9.0

montegoulding commented:
Travis is failing because of the dependency on https://github.com/livecode/livecode-thirdparty/pull/119

montegoulding commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -703,6 +703,74 @@ on Android and iOS).
 > to crash.
 
+#### Foreign Code Libraries
+
```
This all should be (some of it already is) in ide/docs/guides/Extending Livecode.md

montegoulding commented:
Rebased on develop-9.0. I created a new patch for thirdparty due to significant changes causing conflict rebasing it was simpler to just do it again than fix conflicts. https://github.com/livecode/livecode-thirdparty/pull/120

montegoulding commented:
This patch updates `Extending LiveCode.md` https://github.com/livecode/livecode-ide/pull/1995

runrevmark commented on libscript/include/libscript/script.h:
```diff
@@ -27,6 +27,8 @@
 #include <foundation-system.h>
 #endif
 
+#include <ffi.h>
```
This causes a build error on windows - I suggest not having the include here (as libffi should be private to libscript), and replace ffi_abi with its typedef'd type :)

montegoulding replied:
hmm... ok, I won't need the patch to thirdparty then which resolved the build error

montegoulding commented:
This no longer requires a patch to thirdparty

montegoulding commented on tests/lcb/compiler/frontend/variadic.compilertest:
```diff
@@ -17,7 +17,7 @@
 %TEST VariadicLastInForeignHandler
 module compiler_test
-foreign handler TestVariadic(in pA as any, ...) returns nothing binds to ""
+foreign handler TestVariadic(in pA as any, ...) returns nothing binds to "dummy"
```
@runrevmark this is now required because we now get parser errors on binding strings... aka `MCScriptThrowMissingFunctionInForeignBindingError`

montegoulding committed:
[[ lc-compile ]] Update builtin types for shim generation

montegoulding committed:
[[ lc-compile ]] Add --forcebuiltins option

This new option allows any C foreign handler to be generated as a builtin shim
when output is in `--outputauxc` mode. This option also sets the module initializer
and module finalizer functions to nullptr.

montegoulding committed:
[[ libscript ]] Refactor foreign handler binding parsing for reuse

montegoulding committed:
[[ Module Lcext ]] Allow builtin module ensuer usable to fail

This patch changes `MCScriptInitalize` to allow `MCScriptEnsureModuleIsUsable` to fail as
some dependencies may be in deploy capsules.

montegoulding committed:
[[ Module Lcext ]] Allow module section to be a module name to resolve

This patch changes the deploy capsule for module files to permit either the data
of a module file or the name of a module to lookup. If it is a name then the module
should be included in the standalone as a builtin.

montegoulding committed:
[[ Module Lcext ]] Implement lcext linking with auxc output script

montegoulding committed:
[[ libscript ]] Export MCScriptRegisterBuiltinModule
